users/vieting/experiments/librispeech/librispeech_100_ctc/fairseq_finetuning/ctc_standalone/pytorch_networks/wav2vec/w2v_ctc_wrapper.py

```diff
@@ -0,0 +1,137 @@
+import numpy as np
+import torch
+from torch import nn
+
+import fairseq
+from fairseq.dataclass import FairseqDataclass
+from fairseq.models.wav2vec.wav2vec2_asr import Wav2VecCtc, Wav2Vec2CtcConfig
+
+from .w2v_ctc_wrapper_cfg import ModelConfig
+
+class Model(torch.nn.Module):
+    def __init__(self, w2v_config_updates, **kwargs):
+        """
+        Args:
+            w2v_config_updates: dictionary containing the configuration updates for the default FairseqConfig
+        """
+        super().__init__()
+        self.model_config = ModelConfig.from_dict(w2v_config_updates)
+        self.fairseq_config = self.model_config.build_full_config()
+        self.num_update = 0
+        dummy_task = self.model_config.build_dummy_task() # dummy task to build the model
+        self.w2v_model = Wav2VecCtc.build_model(self.fairseq_config.model, task=dummy_task) 
+
+    def get_num_updates(self):
+        """
+        Get the number of gradient updates that have been performed so far
+        """
+        return self.num_update
+    
+    def set_num_updates(self, num_update):
+        """
+        Set the number of gradient updates that have been performed so far
+        """
+        self.num_update = num_update
+        self.w2v_model.w2v_encoder.set_num_updates(num_update)
```

Comment:
Is this needed for the w2v model to work properly?

Reply:
Yes, this is required in order to freeze the transformer for the first `freeze_finetune_updates` updates automatically. Otherwise, this would be needed to be done manually in our forward function.

Reply:
Yes, this way the w2v_model can automatically freeze the transformer for the first `freeze_finetune_updates` updates. Otherwise, we would need to do this manually.